06_gpu_and_ml/stable_diffusion/comfy_ui.py

```diff
@@ -0,0 +1,148 @@
+        "cd /root && git checkout master",
+        "cd /root && pip install xformers!=0.0.18 -r requirements.txt --extra-index-url https://download.pytorch.org/whl/cu121",
+    )
+    # Use fork until https://github.com/valohai/asgiproxy/pull/11 is merged.
+    .pip_install(
+        "git+https://github.com/modal-labs/asgiproxy.git", "httpx", "tqdm"
+    )
+    .run_function(download_checkpoint)
+)
+stub = modal.Stub(name="example-comfy-ui", image=image)
+
+# ## Spawning ComfyUI in the background
+#
+# Inside the container, we will run the ComfyUI server and execution queue in a background subprocess using
+# `subprocess.Popen`. Here we define `spawn_comfyui_in_background()` to do this and then poll until the server
+# is ready to accept connections.
+
+HOST = "127.0.0.1"
+PORT = "8188"
+
+
+def spawn_comfyui_in_background():
+    import socket
+    import subprocess
+
+    process = subprocess.Popen(
+        [
+            "python",
+            "main.py",
+            "--dont-print-server",
+            "--port",
+            PORT,
+        ]
+    )
+
+    # Poll until webserver accepts connections before running inputs.
+    while True:
+        try:
+            socket.create_connection((HOST, int(PORT)), timeout=1).close()
+            print("ComfyUI webserver ready!")
+            break
+        except (socket.timeout, ConnectionRefusedError):
+            # Check if launcher webserving process has exited.
+            # If so, a connection can never be made.
+            retcode = process.poll()
+            if retcode is not None:
+                raise RuntimeError(
+                    f"comfyui main.py exited unexpectedly with code {retcode}"
+                )
+
+
+# ## Wrap it in an ASGI app
+#
+# Finally, Modal can only serve apps that speak the [ASGI](https://modal.com/docs/guide/webhooks#asgi) or
+# [WSGI](https://modal.com/docs/guide/webhooks#wsgi) protocols. Since the ComfyUI server uses `aiohttp`,
+# which [does not support either](https://github.com/aio-libs/aiohttp/issues/2902), we run a separate ASGI
+# app using the `asgiproxy` package that proxies requests to the ComfyUI server.
+
+
+@stub.function(
+    gpu="any",
+    # Allows 100 concurrent requests per container.
+    allow_concurrent_inputs=100,
+    # Restrict to 1 container because we want to our ComfyUI session state
+    # to be on a single container.
+    concurrency_limit=1,
+    timeout=10 * 60,
+)
+@modal.asgi_app()
+def web():
+    from asgiproxy.config import BaseURLProxyConfigMixin, ProxyConfig
+    from asgiproxy.context import ProxyContext
+    from asgiproxy.simple_proxy import make_simple_proxy_app
+
+    spawn_comfyui_in_background()
+
+    config = type(
+        "Config",
+        (BaseURLProxyConfigMixin, ProxyConfig),
+        {
+            "upstream_base_url": f"http://{HOST}:{PORT}",
+            "rewrite_host_header": f"{HOST}:{PORT}",
+        },
+    )()
```

Comment:
hmm I think we should try to avoid repeating this `type()` magic

Reply:
Yeh I just copy-pasted this from the Streamlit example. I don't have the background on why it was setup this way.